### ISSUE TITLE
LogData payload data is copied 3 times before sending to log unit server.

### ISSUE THREAD

**Issue by WenbinZhu** [CorfuDB/CorfuDB#1949]
## Overview

1. First time: `checkMaxWriteSize` opens a `SerializationHandle` that calls `acquireBuffer` and do serialization which copies serialized data into `serializedCache`. After `checkMaxWriteSize` is finished, it closes the `SerializationHandle` which release the `serializedCache` and reassign it to NULL

2. Second time: `ChainReplicationProtocol` write() also opens a `SerializationHandle` and call `acquireBuffer`, since `serializedCache` is assigned to NULL at the end of previous step, it will do serialize again, and copies to `serializedCache`.

3. Third time: `ChainReplicationProtocol` calls `LogUnitClient` write(), which constructs  a `WriteRequest`, when serializing the request, it will call `LogData::doSerialize`. This time the `SerializationHandle` is not released yet, so `serializedCache` is still valid, but still needs to be copied to the Netty buffer.

We should reduce the number of copies as this is a considerable performance overhead.

**Comment by Maithem** [comment]
I don't think the payload is serialized multiple times, you can see this simply by adding this

```java

     void doSerializeInternal(ByteBuf buf) {
+        log.info("Serializing1 at {}", getGlobalAddress());
         ICorfuPayload.serialize(buf, type);
         if (type == DataType.DATA) {
             if (data == null) {
```

On a 3 node cluster, this is the output:

3338 [pool-5-thread-1] INFO  org.corfudb.protocols.wireprotocol.LogData  - Serializing1 at 58
3346 [pool-5-thread-1] INFO  org.corfudb.protocols.wireprotocol.LogData  - Serializing1 at 59
3353 [pool-5-thread-1] INFO  org.corfudb.protocols.wireprotocol.LogData  - Serializing1 at 60
3361 [pool-5-thread-1] INFO  org.corfudb.protocols.wireprotocol.LogData  - Serializing1 at 61
3371 [pool-5-thread-1] INFO  org.corfudb.protocols.wireprotocol.LogData  - Serializing1 at 62
3380 [pool-5-thread-1] INFO  org.corfudb.protocols.wireprotocol.LogData  - Serializing1 at 63

**Comment by WenbinZhu** [comment]
@Maithem I just reproduced with a log inserted in `doSerializeInternal`(same as your example). Based on my test code, it is calling `doSerializeInternal` twice:

```
2019-08-07T05:42:44.713Z | INFO  |            test-main | ocols.wireprotocol.LogData.doSerializeInternal:230 | doSerializeInternal: serializing1 at -1
2019-08-07T05:42:44.795Z | INFO  |            test-main | ocols.wireprotocol.LogData.doSerializeInternal:230 | doSerializeInternal: serializing1 at 0

2019-08-07T05:42:45.235Z | INFO  |            test-main | ocols.wireprotocol.LogData.doSerializeInternal:230 | doSerializeInternal: serializing1 at -1
2019-08-07T05:42:45.239Z | INFO  |            test-main | ocols.wireprotocol.LogData.doSerializeInternal:230 | doSerializeInternal: serializing1 at 1
```

I put two entries into the table, notice the -1 address, this is because the first time we call `doSerializeInternal` is from `checkMaxWriteSize`, by that time, the token is not assigned, so the address is still -1. This shows that `doSerializeInternal` is called twice. And the third copy I mentioned in the issue is copying from `serializedCache` to Netty buffer.

**Comment by WenbinZhu** [comment]
@Maithem Can you check if you could see -1 address in your test output?

**Comment by Maithem** [comment]
I see, it serializes the payload twice. Third step is actually really cheap since its just a buffer copy. Netty will not end up calling doSerializeInternal. The problem is we can't generate a serialization handle once and use it for both the size check and to replication. The reason is, we want to do the size check before the token acquisition step to prevent spinning threads of generating so many holes. 

I'm working on a change to update the token even after serialization, hopefully we can get rid of the second serialization call.

**Comment by WenbinZhu** [comment]
Addressed in #2703
